### PR TITLE
[CM_BE_29] Make temp token usable only once and against valid scope.

### PR DESCRIPTION
**Pull request by ajayvk-tw**

### CODE REVIEW

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestController.java [hunk 1]
```diff
@@ -75,9 +80,14 @@ private static RequestCreatedRepresentation buildResponse(String requestId) {
...
-                .flatMap(username ->
-                        consentManager.approveConsent(username, requestId, consentApprovalRequest.getConsents()));
+                .flatMap(caller ->
+                        consentManager.
+                                approveConsent(caller.getUsername(), requestId, consentApprovalRequest.getConsents())
+                                .flatMap(response -> {
+                                    logger.debug("[approve] putting {} in used tokens",caller.getSessionId());
+                                    return usedTokens.put(caller.getSessionId(),"").thenReturn(response);
+                                })
+                );
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestController.java:90]:
```suggestion
                                }));
```

File: src/main/resources/static/api.yaml [hunk 2]
```diff
@@ -2056,16 +2056,21 @@ components:
...
       type: object
       properties:
         pin:
           type: string
           maximum: 9999
           example: "1234"
           description: '4 digit number'
+        scope:
+          type: string
+          enum : [consentrequest.approve, consentrequest.revoke]
```
Comment (ganesanarun) [inline src/main/resources/static/api.yaml:2069]:
```suggestion
          enum : [consentrequest.approve, consent.revoke]
```

File: src/main/java/in/projecteka/consentmanager/user/TransactionPinService.java [hunk 3]
```diff
@@ -58,16 +62,20 @@ private boolean isPinValid(String pin) {
...
-                }).thenReturn(newToken(patientId));
+                }).thenReturn(newToken(patientId, scope));
     }
 
     @SneakyThrows
-    private Token newToken(String userName) {
+    private Token newToken(String userName, String scope) {
         int minutes = userServiceProperties.getTransactionPinTokenValidity() * 60 * 1000;
+        HashMap<String, Object> claims = new HashMap<>(1);
+        claims.put("sid" , UUID.randomUUID().toString());
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/user/TransactionPinService.java:73]: session id introduced to avoid storing the whole token.

Comment (ajayvk-tw) [reply src/main/java/in/projecteka/consentmanager/user/TransactionPinService.java:73]: Yes

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

File: src/main/resources/static/api.yaml (hunk 2 above)
Comment (ajayvk-tw) [reply src/main/resources/static/api.yaml]: Done.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/SecurityConfiguration.java  1
- src/main/java/in/projecteka/consentmanager/consent/PinVerificationTokenService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5413216&bid=17748000)

**Pull request merged**